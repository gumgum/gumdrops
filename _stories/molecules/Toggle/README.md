`<Toggle>` components come in two flavors: radio and checkbox. Each one behaves in the same manner as their respective standard input type.

**NOTE**: This component can be controlled or uncontrolled. This component accepts any prop that you may want to add, which are applied to the input. The only drawback is that there is not validation for unlisted props. **It is highly recommended that you use controlled components in React.**

_Recommended Props_: `name`, `checked`, `defaultChecked`, `onChange`

## Props

The following props can be passed to the `<Toggle>` component. Any additional props will be passed to the `<input>` :

| name      | type     | description                                               |
| --------- | -------- | --------------------------------------------------------- |
| className | `String` | Additional class to be added to the outermost element     |
| label     | `String` | Text displayed as the label for the toggle                |
| offText   | `String` | Label displayed when toggled off                          |
| onText    | `String` | Label displayed when toggled on                           |
| size      | `String` | Indicate the size of the table. One of `xs` or `sm`       |
| type      | `String` | Indicate the type of toggle. One of `checkbox` or `radio` |

**Controlled example - Example for single Toggle**:

```
state = {
    myToggle: false
}

_handleToggle = () => {
    this.setState(({ myToggle }) => ({ myToggle: !myToggle }));
}

return (
    <Toggle
        name="myToggle"
        label="Default Toggle"
        checked={ this.state.myToggle }
        onChange={ this._handleToggle }
        offText="Off"
        onText="On"
    />
)
```

**Uncontrolled example**:

```
_handleSubmit = (e) => {
    e.preventDefault();
    const { myToggle } = e.target.elements;
    console.log(myToggle.checked);
};

render() {
    return (
        <form onSubmit={ this._handleSubmit }>
            <FormGroup>
                { /* Pass defaultChecked if you want the input to be checked initially */ }
                <Toggle
                    name="myToggle"
                    label="Default Toggle"
                    defaultChecked
                />
            </FormGroup>
        </form>
    );
}
```
