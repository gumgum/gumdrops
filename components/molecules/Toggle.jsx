import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

const Toggle = ({ type, size, label, style, className, onText, offText, ...otherProps }) => {
    const baseClass = 'gds-form-group__toggleswitch';

    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${size}`]: size
    });

    return (
        <div className={rootClass} style={style}>
            <label className="gds-form-group__toggleswitch-label">
                <input className="gds-form-group__toggleswitch-input" type={type} {...otherProps} />
                <span className="gds-form-group__toggleswitch-indicator">
                    <span
                        className="gds-form-group__toggleswitch-indicator-labels"
                        gg-enabled-text={onText}
                        gg-disabled-text={offText}
                    />
                </span>
                {label}
            </label>
        </div>
    );
};

Toggle.displayName = 'Toggle';

Toggle.defaultProps = {
    type: 'checkbox',
    label: '',
    onText: '',
    offText: ''
};

Toggle.propTypes = {
    /** radio, checkbox */
    type: PropTypes.oneOf(['radio', 'checkbox']),
    label: PropTypes.string,
    /** xs, sm */
    size: PropTypes.string,
    style: PropTypes.object,
    className: PropTypes.string,
    onText: PropTypes.string,
    offText: PropTypes.string
};

export default Toggle;
