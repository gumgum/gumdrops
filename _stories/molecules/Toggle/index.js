import React from 'react';
import { select, text, object } from '@storybook/addon-knobs';
import { optionalSelect } from '../../../components/utils/optionalSelect';

import readme from './README.md';
import Toggle from '../../../components/molecules/Toggle';
import { action } from '@storybook/addon-actions';

const options = {
    checkbox: 'checkbox',
    radio: 'radio'
};

const sizeOptions = {
    sm: 'sm',
    xs: 'xs',
    'No Value': ''
};

const component = () => (
    <Toggle
        label={text('Label', 'Default Toggle')}
        type={select('Type', options, 'checkbox')}
        size={optionalSelect('Size', sizeOptions, '')}
        className={text('ClassName', '')}
        style={object('Style', {})}
        onChange={action('toggle_checked')}
        onText={text('OnText', '')}
        offText={text('OffText', '')}
    />
);

export default [readme, component];
