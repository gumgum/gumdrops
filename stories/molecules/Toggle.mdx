import { Story, Source, Props } from '@storybook/addon-docs/blocks';
import Toggle from '../../components/molecules/Toggle';

# Toggle

`<Toggle>` components come in two flavors: radio and checkbox. Each one behaves in the same manner as their respective standard input type.

## Controlled vs Uncontrolled

This component can be controlled or uncontrolled. This component accepts any prop that you may want to add. The only drawback is that there is not validation for unlisted props. **It is highly recommended that you use controlled components in React.**

_Recommended Props_: `name`, `checked`, `defaultChecked`, `onChange`

### Controlled example

```jsx
const [isToggled, setIsToggled] = useState(false);
return (
    <Toggle
        name="myToggle"
        label="Default Toggle"
        checked={isToggled}
        onChange={() => setIsToggled(!isToggled)}
        offText="Off"
        onText="On"
    />
);
```

### Uncontrolled example

```jsx
_handleSubmit = (e) => {
    e.preventDefault();
    const { myToggle } = e.target.elements;
    console.log(myToggle.checked);
};

render() {
    return (
        <form onSubmit={ this._handleSubmit }>
            <FormGroup>
                { /* Pass defaultChecked if you want the input to be checked initially */ }
                <Toggle
                    name="myToggle"
                    label="Default Toggle"
                    defaultChecked
                />
            </FormGroup>
        </form>
    );
}
```

## Component Example

<Story id="molecules-toggle--default" />

## Source

<Source id="molecules-toggle--default" />

## Props

<Props of={Toggle} />
