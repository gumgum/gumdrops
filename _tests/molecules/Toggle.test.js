import renderer from 'react-test-renderer';
import React from 'react';
import Toggle from '../../components/molecules/Toggle';

const defaultProps = {
    type: 'checkbox',
    label: 'Toggle Label',
    size: 'sm',
    style: { width: 100 },
    className: 'custom-class'
};

test('Expect <Toggle> to render properly', () => {
    const tree = renderer.create(<Toggle {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
});
