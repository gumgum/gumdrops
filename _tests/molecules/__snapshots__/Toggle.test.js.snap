// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Toggle> to render 1`] = `
<Toggle
  className="custom-class"
  label="Toggle Label"
  offText=""
  onText=""
  size="sm"
  style={
    Object {
      "width": 100,
    }
  }
  type="checkbox"
>
  <div
    className="gds-form-group__toggleswitch custom-class gds-form-group__toggleswitch--sm"
    style={
      Object {
        "width": 100,
      }
    }
  >
    <label
      className="gds-form-group__toggleswitch-label"
    >
      <input
        className="gds-form-group__toggleswitch-input"
        type="checkbox"
      />
      <span
        className="gds-form-group__toggleswitch-indicator"
      >
        <span
          className="gds-form-group__toggleswitch-indicator-labels"
          gg-disabled-text=""
          gg-enabled-text=""
        />
      </span>
      Toggle Label
    </label>
  </div>
</Toggle>
`;
